rust/wal3/src/shard_manager.rs

```diff
@@ -0,0 +1,224 @@
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Mutex;
+use std::time::{Duration, Instant};
+
+use crate::{
+    DeltaSeqNo, Error, LogPosition, Manifest, ManifestManager, Message, ShardID, ShardSeqNo,
+    StreamID, ThrottleOptions, BATCH_SIZE,
+};
+
+//////////////////////////////////////////// ShardState ////////////////////////////////////////////
+
+#[derive(Debug)]
+struct ShardState {
+    shard_id: ShardID,
+    next_seq_no: ShardSeqNo,
+    next_write: Instant,
+    writers_active: usize,
+}
+
+impl ShardState {
+    fn set_next_write(&mut self, options: &ThrottleOptions) {
+        self.next_seq_no += 1;
+        self.next_write =
+            Instant::now() + Duration::from_micros(1_000_000 / options.throughput as u64);
+    }
+}
+
+/////////////////////////////////////////// ManagerState ///////////////////////////////////////////
+
+#[derive(Debug)]
+#[allow(clippy::type_complexity)]
+struct ManagerState {
+    shards: Vec<ShardState>,
+    enqueued: Vec<(
+        StreamID,
+        Message,
+        tokio::sync::oneshot::Sender<Result<LogPosition, Error>>,
+    )>,
+    last_batch: Instant,
+}
+
+impl ManagerState {
+    fn select_shard_for_write(
+        &mut self,
+        options: &ThrottleOptions,
+        manifest_manager: &ManifestManager,
+        record_count: usize,
+    ) -> Option<(ShardID, ShardSeqNo, LogPosition, DeltaSeqNo)> {
+        let shard_state = self
+            .shards
+            .iter_mut()
+            .filter(|x| x.writers_active < options.outstanding)
+            .min_by_key(|x| x.next_write)?;
+        if shard_state.next_write > Instant::now() {
+            return None;
+        }
+        let (log_position, delta_seq_no) = match manifest_manager.assign_timestamp(record_count) {
+            Some(log_position) => log_position,
+            None => {
+                panic!("log full");
```

Comment:
If we think its possible to have log position exceeding `u64::MAX`, then we should handle this properly instead of `panic`. If it is not supposed to happen, then `assign_timestamp` should not return an `Option<T>`. Either way, `panic` is not really the best option here.